vllm/v1/executor/multiproc_executor.py

```diff
@@ -0,0 +1,197 @@
+            os.environ["VLLM_WORKER_MULTIPROC_METHOD"] = "spawn"
+
+        # Ensure that VLLM_INSTANCE_ID is set, to be inherited by workers
+        os.environ["VLLM_INSTANCE_ID"] = get_vllm_instance_id()
+
+        # Disable torch async compiling which won't work with daemonic processes
+        os.environ["TORCHINDUCTOR_COMPILE_THREADS"] = "1"
+
+        # Configure thread parallelism if OMP_NUM_THREADS isn't set
+        #
+        # Helps to avoid CPU contention. The default of spawning a thread per
+        # core combined with multiprocessing for each GPU can have a negative
+        # impact on performance. The contention is amplified when running in a
+        # container where CPU limits can cause throttling.
+        default_omp_num_threads = 1
+        if "OMP_NUM_THREADS" not in os.environ and (
+                current_parallelism :=
+                torch.get_num_threads()) > default_omp_num_threads:
+            logger.warning(
+                "Reducing Torch parallelism from %d threads to %d to avoid "
+                "unnecessary CPU contention. Set OMP_NUM_THREADS in the "
+                "external environment to tune this value as needed.",
+                current_parallelism, default_omp_num_threads)
+            os.environ["OMP_NUM_THREADS"] = str(default_omp_num_threads)
+            torch.set_num_threads(default_omp_num_threads)
+
+        # workaround for https://github.com/vllm-project/vllm/issues/6103
+        if world_size > 1:
+            maybe_set_triton_cache_manager()
+
+        # Multiprocessing-based executor does not support multi-node setting.
+        # Since it only works for single node, we can use the loopback address
+        # 127.0.0.1 for communication.
+        distributed_init_method = get_distributed_init_method(
+            "127.0.0.1", get_open_port())
+
+        # Initialize worker and set up message queues for SchedulerOutputs
+        # and ModelRunnerOutputs
+        self.worker_request_mq = MessageQueue(world_size, world_size)
+        scheduler_output_handle = self.worker_request_mq.export_handle()
+
+        # Create workers
+        self.workers: List[WorkerProcHandle] = []
+        for rank in range(world_size):
+            worker = WorkerProc.make_worker_process(vllm_config, rank, rank,
+                                                    distributed_init_method,
+                                                    scheduler_output_handle)
+            self.workers.append(worker)
+
+        model_output_mq_handle = self.workers[0].model_output_mq_handle
+        self.model_output_mq = MessageQueue.create_from_handle(
+            model_output_mq_handle, 0)
+
+    def _run_on_workers(self, fn: str, *args) -> List:
+        with ThreadPoolExecutor() as executor:
+            futures = [
+                executor.submit(getattr(type(w), fn), w, *args)
+                for w in self.workers
+            ]
+            result = [f.result() for f in futures]  # Wait for all to complete
```

Comment:
is it possible to directly send the arguments via the shared memory broadcast? here we are serializing the `*args` for every worker i think.

Reply:
This is only used during worker initialization, so I'm not too worried about performance here. I'll add a comment making this clear.

We could use the shared memory broadcast for IPC during initialization, but the way I've designed the PR is that `WorkerProcHandle` is responsible for all the initialization messages, and I'm using ZMQ there because it has fewer gotchas.

Reply:
Just to emphasize: The main difference between this implementation and the one in `multiproc_worker_utils.py` is in the latter, the workers have one flexible event loop that's used for all messages. Here, we have two: one flexible event loop used only for initialization (uses ZMQ), and one very rigid very tight even loop only for model execution (I think we'll need to extend this at least to handle check health)

Reply:
if the flexible event loop is only used for initialization, then we don't need to abstract it to a function?

we should have a general `run_workers` function , it's not just about model execution, but also adding/unloading loras, and maybe other user-defined function.

my design would be to always use the very tight event loop ( shared memory broadcast) for all the function call.

Reply:
Added a general `collective_rpc` function. PTAL!